src/marin/download/uncheatable_eval/download.py

```diff
@@ -0,0 +1,402 @@
+                size=entry.get("size"),
+            )
+        )
+    return datasets
+
+
+def _select_latest_dumps(datasets: Iterable[UncheatableEvalDataset]) -> list[UncheatableEvalDataset]:
+    """Select the latest dump for each benchmark based on the end date (and start date as tie breaker)."""
+
+    latest: dict[str, UncheatableEvalDataset] = {}
+    for dataset in datasets:
+        existing = latest.get(dataset.benchmark)
+        if existing is None:
+            latest[dataset.benchmark] = dataset
+            continue
+        candidate_key = (dataset.end_date, dataset.start_date, dataset.name)
+        existing_key = (existing.end_date, existing.start_date, existing.name)
+        if candidate_key > existing_key:
+            latest[dataset.benchmark] = dataset
+    return sorted(latest.values(), key=lambda d: d.benchmark)
+
+
+def _extract_id(raw: Any, dataset: UncheatableEvalDataset, index: int) -> str:
+    if isinstance(raw, dict):
+        for key in ID_FIELD_CANDIDATES:
+            value = raw.get(key)
+            if value:
+                return str(value)
+        metadata = raw.get("metadata")
+        if isinstance(metadata, dict):
+            for key in ID_FIELD_CANDIDATES:
+                value = metadata.get(key)
+                if value:
+                    return str(value)
+    return f"{dataset.benchmark}_{dataset.date_range}_{index:06d}"
+
+
+def _join_list_field(value: Any) -> str | None:
+    if isinstance(value, list):
+        text_items = [str(item) for item in value if item is not None]
+        if text_items:
+            return "\n".join(text_items)
+    return None
+
+
+def _extract_text(raw: Any) -> str | None:
+    if raw is None:
+        return None
+    if isinstance(raw, str):
+        return raw
+    if isinstance(raw, dict):
+        for key in TEXT_FIELD_CANDIDATES:
+            value = raw.get(key)
+            if isinstance(value, str) and value.strip():
+                return value
+        for key in TEXT_FIELD_CANDIDATES:
+            value = raw.get(key)
+            joined = _join_list_field(value)
+            if joined:
+                return joined
+        for key in LIST_FIELD_CANDIDATES:
+            joined = _join_list_field(raw.get(key))
+            if joined:
+                return joined
+        title = raw.get("title")
+        body = raw.get("body")
+        if isinstance(title, str) and isinstance(body, str):
+            combined = f"{title.strip()}\n\n{body.strip()}"
+            if combined.strip():
+                return combined
+        if isinstance(title, str) and title.strip():
+            return title
+        return json.dumps(raw, ensure_ascii=False)
+    return str(raw)
+
+
+def _normalize_record(raw: Any, dataset: UncheatableEvalDataset, index: int) -> dict[str, str]:
+    text = _extract_text(raw)
+    if text is None or not str(text).strip():
+        raise ValueError(f"Record {index} in {dataset.name} does not contain text")
```

Comment:
It's kind of weird you're passing around the dataset just to make this error message?

That also makes your tests a bit more awkward. You could pass in a "context: str" or "dataset: str" here if you wanted instead.